File: r/sieve/sieve.R
sieve <- function(limit) {
  if (limit < 2) return(c())
  
  is_prime <- rep(TRUE, limit)
  is_prime[1] <- FALSE
  
  for (num in 2:sqrt(limit)) {
    if (is_prime[num] && num^2 < limit) {
      is_prime[seq(num^2, limit, num)] <- FALSE
    }
  }
  which(is_prime)
}

File: JavaScript/sieve/sieve.js
//
// This is only a SKELETON file for the 'Sieve' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const primes = (limit) => {
  if (limit < 2) return [];

  const isPrime = new Array(limit + 1).fill(true);
  isPrime.splice(0, 2, false, false);

  for (let i = 2; i <= Math.sqrt(limit); i++) {
    if (isPrime[i]) {
      for (let j = i ** 2; j <= limit; j += i) {
        isPrime[j] = false;
      }
    }
  }

  return isPrime.reduce(
    (primesList, current, idx) => (current ? [...primesList, idx] : primesList),
    []
  );
};
